src/mbgl/mtl/vertex_attribute.cpp

```diff
@@ -4,78 +4,303 @@
+                get<uint32_t>(value);
+                break;
+            case gfx::AttributeDataType::Float2:
+                get<float>(value);
+                break;
+            default:
+                assert(false);
+                buffer += stride;
+                break;
+        }
     }
-}
+
+    void operator()(VertexAttribute::float3 value) {
+        switch (dataType) {
+            case gfx::AttributeDataType::Byte3:
+                get<int8_t>(value);
+                break;
+            case gfx::AttributeDataType::UByte3:
+                get<uint8_t>(value);
+                break;
+            case gfx::AttributeDataType::Short3:
+                get<int16_t>(value);
+                break;
+            case gfx::AttributeDataType::UShort3:
+                get<uint16_t>(value);
+                break;
+            case gfx::AttributeDataType::Int3:
+                get<int32_t>(value);
+                break;
+            case gfx::AttributeDataType::UInt3:
+                get<uint32_t>(value);
+                break;
+            case gfx::AttributeDataType::Float3:
+                get<float>(value);
+                break;
+            default:
+                assert(false);
+                buffer += stride;
+                break;
+        }
+    }
+
+    void operator()(VertexAttribute::float4 value) {
+        switch (dataType) {
+            case gfx::AttributeDataType::Byte4:
+                get<int8_t>(value);
+                break;
+            case gfx::AttributeDataType::UByte4:
+                get<uint8_t>(value);
+                break;
+            case gfx::AttributeDataType::Short4:
+                get<int16_t>(value);
+                break;
+            case gfx::AttributeDataType::UShort4:
+                get<uint16_t>(value);
+                break;
+            case gfx::AttributeDataType::Int4:
+                get<int32_t>(value);
+                break;
+            case gfx::AttributeDataType::UInt4:
+                get<uint32_t>(value);
+                break;
+            case gfx::AttributeDataType::Float4:
+                get<float>(value);
+                break;
+            default:
+                assert(false);
+                buffer += stride;
+                break;
+        }
+    }
+
+    void operator()(VertexAttribute::matf3) {
+        assert(false);
+        buffer += stride;
+    }
+    void operator()(VertexAttribute::matf4) {
+        assert(false);
+        buffer += stride;
+    }
+    void operator()(VertexAttribute::ushort8) {
+        assert(false);
+        buffer += stride;
+    }
+
+private:
+    gfx::AttributeDataType dataType;
+    std::uint8_t* buffer{nullptr};
+    std::size_t stride{0};
+
+    template <typename To, typename From, typename = std::enable_if_t<std::is_assignable_v<To&, From>>>
+    void get(const From value) {
+        assert(stride == sizeof(To));
+        *reinterpret_cast<To*>(buffer) = static_cast<To>(value);
```

Comment:
I know this is just an internal helper class, but this seems asking for trouble. It takes one off by one error or one call to many to operator() to introduce an out of bounds write.

Adding this functionality to `mtl::VertexAttribute` may be a more memory safe alternative to consider.